flux-refineck/src/checker.rs

```diff
@@ -125,67 +133,133 @@ impl<'a, 'tcx, P> Checker<'a, 'tcx, P> {
+    }
+
+    fn init(&mut self, def_id: DefId) {
+        self.inner.entry(def_id).or_insert_with(InferResult1::new);
+    }
+}
+
+impl InferResult1 {
+    fn new() -> Self {
+        Self { inner: FxHashMap::default() }
+    }
+
+    fn index(&self, bb: &BasicBlock) -> &TypeEnvInfer {
+        &self.inner[bb]
+    }
+
+    fn get(&self, bb: &BasicBlock) -> Option<&TypeEnvInfer> {
+        self.inner.get(bb)
+    }
+
+    fn entry(&mut self, bb: &BasicBlock) -> Entry<rustc_middle::mir::BasicBlock, TypeEnvInfer> {
+        self.inner.entry(*bb)
+    }
+}
+
+impl<'a, 'tcx> Checker<'a, 'tcx, Inference> {
     pub fn infer(
         genv: &GlobalEnv<'a, 'tcx>,
         body: &Body<'tcx>,
         def_id: DefId,
-    ) -> Result<FxHashMap<BasicBlock, TypeEnvInfer>, CheckerError> {
+    ) -> Result<InferResult, CheckerError> {
         dbg::infer_span!(genv.tcx, def_id).in_scope(|| {
-            let mut bb_envs = FxHashMap::default();
+            // let mut bb_envs = FxHashMap::default();
+            // let mut bb_envs = InferResult::new();
+            let mut phase = Inference { bb_envs: InferResult::new() };
+
+            let fn_sig = genv.lookup_fn_sig(def_id).unwrap_or_else(|_| {
+                span_bug!(body.span(), "checking function with unsupported signature")
+            });
+
             Checker::run(
                 genv,
                 RefineTree::new().as_subtree(),
-                body,
                 def_id,
-                Inference { bb_envs: &mut bb_envs },
+                &mut phase, // Inference { bb_envs: &mut bb_envs },
+                fn_sig,
             )?;
 
-            Ok(bb_envs)
+            Ok(phase.bb_envs)
         })
     }
 }
 
-impl<'a, 'tcx> Checker<'a, 'tcx, Check<'_>> {
+impl<'a, 'tcx> Checker<'a, 'tcx, Check> {
     pub(crate) fn check(
         genv: &GlobalEnv<'a, 'tcx>,
         body: &Body<'tcx>,
         def_id: DefId,
         refine_tree: RefineSubtree,
-        kvars: &mut KVarStore,
-        bb_envs_infer: FxHashMap<BasicBlock, TypeEnvInfer>,
-    ) -> Result<(), CheckerError> {
-        let bb_envs = bb_envs_infer
-            .into_iter()
-            .map(|(bb, bb_env_infer)| (bb, bb_env_infer.into_bb_env(kvars)))
-            .collect();
-
-        dbg::check_span!(genv.tcx, def_id, bb_envs)
-            .in_scope(|| Checker::run(genv, refine_tree, body, def_id, Check { bb_envs, kvars }))
+        kvars: KVarStore,
+        bb_envs_infer: InferResult,
+    ) -> Result<KVarStore, CheckerError> {
+        let fn_sig = genv.lookup_fn_sig(def_id).unwrap_or_else(|_| {
+            span_bug!(body.span(), "checking function with unsupported signature")
+        });
+        let mut phase = Check { bb_envs_infer, bb_envs: FxHashMap::default(), kvars };
+
+        // TODO(CLOSURE): I broke this check_span thing, help!
+        dbg::check_span!(genv.tcx, def_id /* , bb_envs */)
+            .in_scope(|| Checker::run(genv, refine_tree, def_id, &mut phase, fn_sig))?;
```

Comment:
This breaks [this script](https://github.com/flux-rs/flux/blob/e3bfb5a257b26bfdb2185ec7bf2437168e433388/tools/logreader.py), but we can fix it later.